README
Emacs-Rep

A tool for interactive finds and replaces from inside of emacs,
but using perl5 substitution commands.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

Additionally, you need to make sure that this file has
been added to a location in your emacs load-path:

        elisp/rep.el

And as explained at the top of that file:

   Add the following into your ~/.emacs (or equivalent):
     (require 'rep)
     (global-set-key "C-c.S" 'rep-open-substitutions)

SUPPORT AND DOCUMENTATION

User documentation can be found in the comments at the
top of:

   elisp/rep.el

Web pages related to this can be found at:
  http://obsidianrook.com/rep

After installing, you can find documentation for this module with
the perldoc command.

    perldoc Emacs::Rep

A good discussion forum for perl/emacs lash-ups like this:
  http://groups.google.com/group/emacs-perl-intersection

The code is available on github:
  http://github.com/doomvox/rep

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Emacs-Rep

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Emacs-Rep

    CPAN Ratings
        http://cpanratings.perl.org/d/Emacs-Rep

    Search CPAN
        http://search.cpan.org/dist/Emacs-Rep/


COPYRIGHT AND LICENCE

Copyright (C) 2010 Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Emacs-Rep

0.01    Sun May 30 12:23:55 2010
        First version.
        Limitation: all of perl's substitutions syntax is
        not supported, in particular substitutions with embedded
        semicolons may behave oddly.

0.02    Sun Jun  6 17:36:22 2010
        Using a PPI-based solution now.  The full range of perl's
        substitution syntax is now supported.
        Have support for multi-line substitutions.
        Window management is more careful (no crashes
        from the current window being too small).
        User set-up is more flexible (easy to use an alternative key prefix).

0.03    Fri Jun 11 04:31:04 2010
        Implemented an entirely new system of managing change
        metadata, which now supports an indefinite series of
        individual undos of cascading, overlapping substitution
        changes.
        The modified buffer is now made read-only until changes
        are accepted.
        User set-up with rep-standard-setup is still more flexible.

0.04    Sat Jun 12 22:34:28 2010

        Cleaner handling of zero-width end results (i.e.  when a
        s///g is used to delete a string entirely, now you can
        still tab to that location and undo the deletion).
        Also made modifications to backtab behavior to work better
        with these "zero width" regions.

0.05    Tue Jun 15 11:44:04 2010

        Cleaner implementation of tab/backtab functions in
        rep-modififed-mode:
        These:
          rep-modified-skip-to-next-change and rep-modified-skip-to-prev-change
        now use the primitives:
           rep-at-start-of-changed-region
           rep-at-end-of-changed-region
        which are built on top of the (successful) experiment:
           rep-rising-or-falling-edge



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Emacs::Rep',
    license             => 'perl',
    dist_author         => q{Joseph Brenner <doom@kzsu.stanford.edu>},
    dist_version_from   => 'lib/Emacs/Rep.pm',
    build_requires => {
        'Data::Dumper::Perltidy' => 0,
        'Test::More'             => 0,
        'FindBin'                => 0,
        'lib'                    => 0,
        'Test::Trap'             => 0,
      },
    requires => {
        'PPI'            => 0,
        'strict'         => 0,
        'warnings'       => 0,
        'Carp'           => 0,
        'Data::Dumper'   => 0,
        'Exporter'       => 0,
      },
    add_to_cleanup      => [ 'Emacs-Rep-*' ],
    create_makefile_pl => 'traditional',
    script_files => ['script/rep.pl'],
);

$builder->create_build_script();


script/rep.pl
#!/usr/bin/perl
# rep.pl                   doom@kzsu.stanford.edu
#                          13 May 2010

=head1 NAME

rep.pl - perform a series of find an replaces

=head1 SYNOPSIS

  perl rep.pl --backup <backup_file> --substitutions <filename> --target <file_to_act_on>

  perl rep.pl -b <backup_file> -f <file_to_act_on> 's/foo/bar/'

=head2 USAGE

  rep.pl -[options] [arguments]

  Options:

     -s                substitutions list file
     --substitutions   same
     -f                target file name to be modified
     --target          same
     -B                backup file name
     --backup          same

     -d                debug messages on
     --debug           same
     -h                help (show usage)
     -v                show version
     --version         show version


=head1 DESCRIPTION

B<rep.pl> is a script which does finds and replaces on a file,
and records the beginning and end points of the modified
strings.

It is intended to act as an intermediary between Emacs::Rep
and the emacs lisp code which drives the "rep" process.

Emacs can then use the recorded locations to highlight the
changed regions, and it can use information about what was
replaced to perform undo operations.

The elisp code must choose a unique backup file name. This makes
it possible to do reverts of an entire run of substitutions.

The script returns a serialized data dump of the history
of the changes to the text.  See the documentation routine
L<serialize_change_metadata> in L<Emacs::Rep> for full
details of this output format.

Roughly, you can expect output that looks like:

  0:303:308:1:cars;
  1:113:123:8:of;
  1:431:441:8:of;
  1:596:606:8:of;
  2:330:355:23:.;
  3:702:711:0:evening;
  4:855:863:4:cane;

Where the colon separated fields are:

 first: the "pass" through the file (one pass per substitution command)
 second: begin point of changed string
 third:  end point of changed string
 fourth: the delta, the change in string length
 fifth: the original string that was replaced

Characters are counted from the beginning of the text,
starting with 1.

The fifth field may contain colons, but semicolons should be
escaped with a backslash.

=cut

use warnings;
use strict;
$|=1;
use Carp;
use Data::Dumper;

use File::Path     qw( mkpath );
use File::Basename qw( fileparse basename dirname );
use File::Copy     qw( copy move );
use Fatal          qw( open close mkpath copy move );
use Cwd            qw( cwd abs_path );
use Env            qw( HOME );
use Getopt::Long   qw( :config no_ignore_case bundling );
use FindBin qw( $Bin );
use lib ("$Bin/../lib",
         "$HOME/End/Cave/Rep/Wall/Emacs-Rep/scripts/../lib"); # TODO
use Emacs::Rep     qw( :all );

our $VERSION = 0.06;
my  $prog    = basename($0);

my $DEBUG   = 0;                 # TODO set default to 0 when in production
my ( $locs_temp_file, $reps_file, $backup_file, $target_file );
GetOptions ("d|debug"           => \$DEBUG,
            "v|version"         => sub{ say_version(); },
            "h|?|help"          => sub{ say_usage();   },
            "s|substitutions=s" => \$reps_file,
            "b|backup=s"        => \$backup_file,
            "f|target=s"        => \$target_file,
           ) or say_usage();

# get a series of finds and replaces
#   either from the substitutions file,
#   or from command-line (a series of strings from @ARGV),

my $reps_text;
if( $reps_file ) {
  undef $/;
  open my $fh, '<', $reps_file or croak "$!";
  $reps_text = <$fh>;
} else {
  $reps_text = join( "\n", @ARGV );
}

# process the find_and_reps into an array of find and replace
# pairs (modifiers get moved inside the find.)
my $find_replaces_aref;
eval {
  $find_replaces_aref =
    parse_perl_substitutions( \$reps_text );
};
if ($@) {
  carp "Problem parsing perl substitutions: $@";
  exit;
}

rename( $target_file, $backup_file ) or
  croak "can't copy $target_file to $backup_file: $!";

my $text;
{ undef $/;
  open my $fin, '<', $backup_file or croak "$!";
  $text = <$fin>;
  close( $fin );
}

# Apply the finds and replaces to text, recording the
# change meta-data
my $locations_aref;
eval {
  $locations_aref =
    do_finds_and_reps( \$text, $find_replaces_aref );
};
if ($@) {
  carp "Problem applying finds and replaces: $@";
  rename( $backup_file, $target_file ); # rollback!
} else {
  open my $fout, '>', $target_file or croak "$!";
  print {$fout} $text;
  close($fout);

  # serialize the data to pass to emacs
  my $flat_locs = serialize_change_metadata( $locations_aref );
  print $flat_locs;
}

### end main, into the subs

sub say_usage {
  my $usage=<<"USEME";
   $prog <-options> <arguments>

  Options:

     -s                substitutions list file
     --substitutions   same
     -f                target file name to be modified
     --target          same
     -b                backup file name
     --backup          same

     -d                debug messages on
     --debug           same
     -h                help (show usage)
     -v                show version
     --version         show version

Typical use:

  perl rep.pl -b "/tmp/edit_this.txt.bak" -f "edit_this.txt" 's/foo/bar/'

  perl rep.pl --backup <backup_file> --substitutions <filename> --target <file_to_act_on>

USEME
  print "$usage\n";
  exit;
}

sub say_version {
  print "Running $prog version: $VERSION\n";
  exit 1;
}


__END__

=head1 AUTHOR

Joseph Brenner, E<lt>doom@kzsu.stanford.eduE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 BUGS

None reported... yet.

=cut


elisp/examples/rep-setup.el
;;; rep-setup.el ---

;; Copyright 2010 Joseph Brenner
;;
;; Author: doom@kzsu.stanford.edu
;; Version: $Id: rep-setup.el,v 0.0 2010/05/29 01:17:56 doom Exp $
;; Keywords:
;; X-URL: not distributed yet

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

;;; Commentary:

;;  This is a simple example of how one might do set-up for
;;  the rep.el package.

;;  TODO add more bells and whistles: no point in publishing this
;;  as it stands.

;; Put this file into your load-path and the following into your ~/.emacs:
;;   (require 'rep-setup)

;;; Code:

(provide 'rep-setup)
(eval-when-compile
  (require 'cl))




(require 'rep)
(global-set-key "\C-c.S" 'rep-open-substitutions)

;;; rep-setup.el ends here


lib/Emacs/Rep.pm
package Emacs::Rep;
#                                doom@kzsu.stanford.edu
#                                15 May 2010

=head1 NAME

Emacs::Rep - find & replace backend for rep.pl and in-turn rep.el

=head1 SYNOPSIS

  use Emacs::Rep qw( do_finds_and_reps  parse_perl_substitutions );

   my $substitutions =>>'END_S';
      s/jerk/iconoclast/
      s/conniving/shrewd/
      s/(t)asteless/$1alented/i
  END_S

  my $find_replaces_aref =
    parse_perl_substitutions( \$substitutions );

  my $locations_aref =
        do_finds_and_reps( \$text, $find_replaces_aref );


=head1 DESCRIPTION

Emacs::Rep is a module that acts as a back-end for the
rep.pl script which in turn is used by the emacs library.
rep.el.

It's purpose is to perform multiple perl substitution
commands (e.g. s///g) on a given file, using emacs to
interactively display and control the changes.

The end user isn't expected to need to use these routines
(or even the rep.pl script) directly.

An application programmer might use these to add support
for some other interactive front-end.

=head2 EXPORT

None by default.  Any of the following may be requested (or all
with the ':all' tag).

=over

=cut

use 5.008;
use strict;
use warnings;
my $DEBUG = 1;
use Carp;
use Data::Dumper::Perltidy;
use PPI;

require Exporter;

our @ISA = qw(Exporter);
our %EXPORT_TAGS = ( 'all' => [
  qw(
      do_finds_and_reps
      split_perl_substitutions
      parse_perl_substitutions
      serialize_change_metadata
      revise_locations

      strip_brackets
      accumulate_find_reps
      define_nonbracketed_s_scraper_pat
      split_perl_substitutions
      dequote

    ) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw(  );
our $VERSION = '0.06';  # TODO manually sync-up rep.pl and rep.el versions

=item do_finds_and_reps

Does a series of finds and replaces on some text and
returns the beginning and end points of each of the
modfied regions, along with some other information about
the matches.

Takes two arguments:

(1) A *reference* to the text to be modified.
(2) A series of find and replace pairs in the form
    of an aref of arefs, e.g.

  $find_replaces_aref =
   [ ['jerk',            'iconoclast'],
     ['conniving',       'shrewd'].
     ['(?i)(t)asteless', '$1alented'].
   ]:

Example usage:

$locations_aref =
   do_finds_and_reps( \$text, $find_replaces_aref );

The returned history is an aref of aref of arefs, e.g.

  [
    [
     [ 219,  229, 6, 'jerk'],
     [ 649,  659, 6, 'jerk'],
     [1978, 1988, 6, 'jerk'],
   ],
   [
     [ 402,  408, -3, 'conniving'],
     [1620, 1626, -3, 'conniving'],
   ],
   [
     [ 110,  118, -1, 'tasteless'],
     [ 494,  502, -1, 'tasteless'],
     [1672, 1680, -1, 'tasteless'],
     [1697, 1705, -1, 'Tasteless'],
   ],
  ]

The sub-arrays are records of changes made by a substitution.
The change meta-data included is, in order: the beginning and
end points, the delta (change in length), and the original
string.

The start and end points are in the form of integers counting
from the start of the file, where the first character is 1.

Note that the change locations are recorded *during* each pass,
which means that later passes throw off the numbering, but we
compensate for this internally using the recorded deltas.
See L<revise_locations>.

=cut

sub do_finds_and_reps {
  my $text_ref      = shift;
  my $find_replaces = shift; # aref of aref: a series of pairs

  my $opts = shift;
  my $LIVE_DANGEROUSLY = $opts->{ LIVE_DANGEROUSLY };

  my $text_copy;
  unless( $LIVE_DANGEROUSLY ) {
    $text_copy = ${ $text_ref };
  }

  my @locations;
  my $pass = 0;
  eval {
    foreach my $sub_pair ( @{ $find_replaces } ) {
      my ($find_pat, $replace) = @{ $sub_pair };

      my @pass = ();
      my $delta_sum = 0; # running total of deltas for the pass

      ${ $text_ref } =~
        s{$find_pat}
         {
           my $s = eval "return qq{$replace}"; # TODO no need for return? # eval qq{ $replace };
#           my $s = eval qq{ $replace }; # but just using this breaks 02-*.t
           # ($DEBUG) && print STDERR "match: $&, 1st: $1, subst: $s\n";
           my $l1 = length( $& );
           my $l2 = length( $s );
           my $delta = $l2 - $l1;
           # in here, pos points at the start of the match, and it uses
           # char numbering fixed at the start of the s///ge run
           my $p = pos( ${ $text_ref } ) + 1 + $delta_sum;
           my $beg = $p;
           my $end = $p + $l2;
           push @pass, [$beg, $end, $delta, $&];
           $delta_sum += $delta;
           $s
         }ge;

      push @locations, \@pass;
      $pass++;
    }
  };
  if ($@) {
    # Send error message to STDOUT so that it won't mess up test output.
    # (and anyway, the elisp call shell-command-to-string merges in STDERR)
    #
    # The elisp function rep-run-perl-substitutions uses prefix "Problem".
    # to spot error messages
    print "Problem: $@\n";
    # roll-back
    @locations = ();
    unless( $LIVE_DANGEROUSLY ) {
      ${ $text_ref } = $text_copy;
    }
  }
  revise_locations( \@locations );
  return \@locations; # aref of aref of arefs of pairs
}

=item revise_locations

Example usage (note, revises structure in-place):

  revise_locations( $locs );

This compensates for a problem in the change history recorded
by L<do_finds_and_reps>.

Later passes with another substitution command can move around
the modified strings from previous passes.

This routine does some numerical magic, re-interpreting previous
passes in the light of later ones.

An example of a change history:

 [
  [ [ 3,       9,   -4,  'alpha'],
    [ 39,     47,   10,  'ralpha'],
    [ 111,   130,    0,  'XXX'],
    [ 320,   332,  -33,  'blvd'],
  ],
  [ [ 12,     23,   6,  'widget'],
    [ 33,     80,   6,  'wadget'],
    [ 453,   532,   6,  'wandat'],
  ],
 ]

Given this data, we can see that the first pass needs to be
shifted forward by a delta of 6, acting at the end-point of each
changed region.

So any locations after 23 need to have 6 added to them (and
locations after 80 need another 6 and ones after 532 -- if there
were any -- would need another 6).

=cut

# So: we crunch through the data in reverse order,
# and record accumulated deltas, keyed by the location to apply
# them, i.e. to revise the beg  and end points.
# The size of earlier deltas is untouched, but the position
# the earlier deltas act upon is the revised position.

sub revise_locations {
  my $locs = shift;
  # named array indicies for readability
  my ($BEG, $END, $DELTA, $ORIG ) = 0 .. 3;
  my %delta;
  foreach my $pass ( reverse @{ $locs } ) {
    foreach my $row ( @{ $pass } ) {
      foreach my $spot ( sort {$a <=> $b} keys %delta) {
        if ( $row->[ $BEG ] >= $spot ) {
          $row->[ $BEG ] += $delta{ $spot };
        }
        if ( $row->[ $END ] >= $spot  ) {
          $row->[ $END ] += $delta{ $spot };
        }
      } # end foreach spot
    } # end foreach row
    foreach my $row ( @{ $pass } ) {
      { no warnings 'uninitialized';
        $delta{ $row->[ $END ] } += $row->[ $DELTA ];
      }
    } # end foreach row
  } # end foreach pass
}

=item serialize_change_metadata

Serialize the locations data structure into a text form to be
passed to emacs.

The result is a block of text, where each line has four
integers separated by colons, in this order:

  <pass>:<beg>:<end>:<delta>:<orig>;

The fields:

  pass  -- line number of the substitution command that made the change
  beg   -- beginning of the modified string, integer count starting at 1
  end   -- ending of the modified string, integer count starting at 1
  delta -- the change in character length due to the substitution
  orig  -- the original string that was replaced.

The trailine semi-colon in this format allows it to work easily
on strings with embedded newlines.  Any embedded semi-colons,
will be backslash escaped by this routine.

=cut

sub serialize_change_metadata {
  my $locations = shift;
  my $ret = '';

  my $pass_count = 0;
  foreach my $pass ( @{ $locations } ) {
    foreach my $row ( @{ $pass } ) {
      my ($beg, $end, $delta, $orig) = @{ $row };

      # escape any semicolons
      $orig =~ s{;}{\\;}xmsg;

      $ret .= sprintf "%d:%d:%d:%d:%s;\n",
        $pass_count, $beg, $end, $delta, $orig;
    }
    $pass_count++;
   }
  return $ret;
}

=item parse_perl_substitutions

Scrapes some text looking for various forms of perl substitutions
(i.e. "s///;", "s{}{};", etc.).

Returns the find_replaces data structure used by L<do_finds_and_args>.

Takes one argument, a scalar reference to a block of text
containing multiple perl substitution commands.

The bracketed form (e.g. "s{}{}") is supported, even the
(obscure, if not insane) mixed form is supported: "s{}//".

End of line comments (after the closing semicolon) beginning with a "#",
are allowed.

Multi-line substitutions are also allowed.  Embedded comments inside
a /x formatted pattern are not stripped out as you might expect:
rather they're carried along inside the matched pattern.

Example usage:

my $substitutions =>>'END_S';
s/pointy-haired boss/esteemed leader/
s/death spiral/minor adjustment/
END_S

my $find_replaces_aref =
  parse_perl_substitutions( \$substitutions );

Where the returned data should look like:
(( TODO revise! ))

   [ ['pointy-haired boss', 'esteemed leader'],
     ['death spiral',       'minor adjustment'],
   ]

=cut

sub parse_perl_substitutions {
  my $reps_text_ref = shift;
  my $Document = PPI::Document->new( $reps_text_ref );
  my $s_aref = $Document->find('PPI::Token::Regexp::Substitute');
  my @find_reps;
  foreach my $s_obj (@{ $s_aref }) {
    my $find      = $s_obj->get_match_string;
    my $rep       = $s_obj->get_substitute_string;
    my $modifiers = $s_obj->get_modifiers; # href
    my @delims    = $s_obj->get_delimiters;

    my $raw_mods = join '', keys %{ $modifiers };

    my $open_bracket_pat =
      qr{ ( [({[<] )                            # )}]>
        }xms;

    my $find_delims = $delims[0];
    my $rep_delims  = $delims[1];

    if ($find_delims =~ m{ $open_bracket_pat }xms ) {
      strip_brackets( \$find );
      ### TODO on multi-line $find, strip eol comments
    } elsif ($rep_delims =~ m{ $open_bracket_pat }xms ) {
      strip_brackets( \$rep );
    } else {  # is this actually a *good* idea?
      my $find_sep = substr( $find_delims, 0, 1 );
      my $rep_sep  = substr( $find_delims, 0, 1 );
      dequote( \$find, $find_sep );
      dequote( \$rep,  $rep_sep  );
    }

    accumulate_find_reps( \@find_reps, $find, $rep, $raw_mods );
  }
  \@find_reps;
}


=item accumulate_find_reps

Example usage:

 accumulate_find_reps( \@find_reps, $find, $rep, $raw_mods );

=cut

sub accumulate_find_reps {
  my $find_reps_aref = shift;
  my $find           = shift;
  my $rep            = shift;
  my $raw_mods       = shift;

  if ($raw_mods) {
    # The modifiers we care about (screening out spurious g or e or ;)
    my @mods = qw( x m s i );
    my $mods = '';
    foreach my $m (@mods) {
      if ( $raw_mods =~ qr{$m}x ) {
        $mods .= $m;
      }
    }
    # modify $find to incorporate the modifiers internally
    $find = "(?$mods)" . $find if $mods;
  }

  push @{ $find_reps_aref }, [ $find, $rep ];
}



=item strip_brackets

Removes any balanced pair of surrounding bracket characters
from the referenced string.  Returns 1 for success, 0 for failure.

Text::Balanced's extract_bracketed, in it's infinite wisdom
does not extract what's inside the brackets, but instead
includes the brackets in the output.  This is a utility
to deal with this oddity.

Example usage:

 if( strip_brackets( \$string ) ) {
    print "brackets removed, see: $string\n";
 }

=cut

sub strip_brackets {
  my $string_ref = shift;
  my $open_bracket_pat = qr{ ( [\(\{\[\<] ) }xms;                        # )}]>

  my $original = ${ $string_ref };

  my $status;
  if( defined( $string_ref ) ) {
    my $bracket;
    if( ${ $string_ref } =~ s{ ^ \s* $open_bracket_pat \s* }{}xms ) {
      $bracket = $1;
    } else {
      return; # fail
    }

    my $close;
    if ( $bracket eq '(' ) {
      $close = ')';
    } elsif ( $bracket eq '{' ) {
      $close = '}';
    } elsif ( $bracket eq '[' ) {
      $close = ']';
    } elsif ( $bracket eq '<' ) {
      $close = '>';
    } else {
      carp "Unexpected bracket string capture: $bracket";
      return;                   # fail
    }

    my $close_bracket_pat = qr{ \Q$close\E }xms;

    if ( ${ $string_ref } =~ s{  \s*? $close_bracket_pat \s*  $ }{}xms ) {
      $status = 1;
    } else { # no change on failure
      $status = 0;
      ${ $string_ref } = $original;
    }
  }
  return $status;
}



=item dequote

Removes backwhack quoting, but only from the single character
supplied as a second argument.

Operates on a string reference, modifying it in place.

Example usage:

  $find = '\/home\/doom';
  dequote( \$find, '/' );

  # $find now '/home/doom';

(Sometimes it's easier to roll your own that to find someone else's.)
(Sometimes.)

=cut

sub dequote {
  my $string_ref = shift;
  my $sep = shift;
  ${ $string_ref } =~ s{ \\ \Q$sep\E }{$sep}xmsg;
}

1;

=back

=head1 SEE ALSO

This is the back-end for the script rep.pl which in turn
is the back-end for the emacs lisp code rep.el.

If rep.el is not installed, look in the "elisp" sub-directory
of this CPAN package.

A good discussion forum for projects such as this is:

  http://groups.google.com/group/emacs-perl-intersection

Web pages related to this can be found at:

  http://obsidianrook.com/rep

The code is available on github:

  http://github.com/doomvox/rep


=head1 AUTHOR

Joseph Brenner, E<lt>doom@kzsu.stanford.eduE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 BUGS

None reported... yet.

=cut
